### PR TITLE
expression: return error when doing `ResolveIndices`

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

Originally, there're some cases that `ResolveIndices` find out that there're some columns not resolved although the SQL can be executed successfully.
This pr fixes it and let `ResolveIndices` return error.

### What is changed and how it works?

Fix the following cases:
- Projection's `columnPruning` doesn't prune the `_rowid` column correctly
- When it's a double read with keep order case. There'll be an extra `_rowid` column returned and changed the schema of operators above the DataSource operator.
- When push agg down across union, union's schema is not set correctly.

And changed the `ResolveIndices` method let it return an error.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has interface methods change

Side effects

 - Increased code complexity

Related changes

None

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8929)

### CODE REVIEW

File: expression/column.go [hunk 1]
```diff
@@ -313,17 +314,18 @@ func (col *Column) HashCode(_ *stmtctx.StatementContext) []byte {
...
+	err := newCol.resolveIndices(schema)
+	return newCol, err
 }
 
-func (col *Column) resolveIndices(schema *Schema) {
+func (col *Column) resolveIndices(schema *Schema) error {
 	col.Index = schema.ColumnIndex(col)
 	if col.Index == -1 {
-		log.Errorf("Can't find column %s in schema %s", col, schema)
+		return errors.Trace(errors.Errorf("Can't find column %s in schema %s", col, schema))
```
Comment (XuHuaiyu) [inline expression/column.go]: remove errors.Trace?

File: planner/core/rule_column_pruning.go [hunk 2]
```diff
@@ -69,6 +69,11 @@ func (p *LogicalProjection) PruneColumns(parentUsedCols []*expression.Column) {
 			p.Exprs = append(p.Exprs[:i], p.Exprs[i+1:]...)
 		}
 	}
+	for k, cols := range p.schema.TblID2Handle {
```
Comment (XuHuaiyu) [inline planner/core/rule_column_pruning.go]: I think we need to add a comment here.

File: planner/core/task.go [hunk 3]
```diff
@@ -47,6 +47,8 @@ type copTask struct {
 	indexPlanFinished bool
 	// keepOrder indicates if the plan scans data by order.
 	keepOrder bool
+	// In double read case,
```
Comment (XuHuaiyu) [inline planner/core/task.go]: unfinished comment.

File: expression/column.go (hunk 1 above)
Comment (winoros) [reply expression/column.go]: Errorf won't add `StackTree` automatically.
And error stack is helpful for this kind of error.

Comment (zz-jason) [reply expression/column.go]: `errors.Trace()` will be removed in the near future. It's better to generate an error with stack and remove the `errors.Trace()` function call.

File: planner/core/resolve_indices.go [hunk 4]
```diff
@@ -65,206 +72,412 @@ func refine4NeighbourProj(p, childProj *PhysicalProjection) {
...
+		if err != nil {
+			return err
+		}
 	}
+	return err
 }
 
-func (p *physicalSchemaProducer) ResolveIndices() {
-	p.basePhysicalPlan.ResolveIndices()
+/*
```
Comment (zz-jason) [inline planner/core/resolve_indices.go]: can be removed?

File: planner/core/resolve_indices.go [hunk 5]
```diff
@@ -65,206 +72,412 @@ func refine4NeighbourProj(p, childProj *PhysicalProjection) {
...
+func (p *PhysicalUnionAll) ResolveIndices() (err error) {
```
Comment (lamxTyler) [inline planner/core/resolve_indices.go]: Remove it?

File: expression/column.go (hunk 1 above)
Comment (winoros) [reply expression/column.go]: `errors.ErrorfWithStack(format string, args ...interface{})`?
`pingcap/errors` have a method named `WithStack(err)` which i think it's the same thing with `errors.Trace`

**Pull request merged**